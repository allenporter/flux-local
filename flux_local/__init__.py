"""
.. include:: ../README.md
"""

__all__ = [
    "git_repo",
    "manifest",
    "kustomize",
    "helm",
    "exceptions",
]
