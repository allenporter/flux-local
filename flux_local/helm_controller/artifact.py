"""
HelmRelease artifact types.

This module defines the artifact types used by the HelmRelease controller
for storing and managing Helm-related artifacts.
"""

from dataclasses import dataclass
from typing import Any

import yaml

from flux_local.store.artifact import Artifact


@dataclass(frozen=True, kw_only=True)
class HelmReleaseArtifact(Artifact):
    """Artifact representing a HelmRelease build result.

    This is written after a HelmRelease is reconciled to record the
    rendered Kubernetes objects and values used for rendering.
    """

    chart_name: str
    """HelmChart name used to render this release for informational purposes."""

    values: dict[str, Any]
    """Resolved values used for rendering the HelmRelease."""

    objects: list[dict[str, Any]]
    """The rendered Kubernetes objects, output of templating the Helm Chart."""

    @property
    def manifests(self) -> list[str]:
        """List of rendered Kubernetes manifests as YAML strings."""
        return [yaml.dump(obj, sort_keys=False) for obj in self.objects]
