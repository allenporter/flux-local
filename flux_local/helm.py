"""Library for running `helm template` to produce local items in the cluster.

You can instantiate a helm template with the following:
- A HelmRepository which is a url that contains charts
- A HelmRelease which is an instance of a HelmChart in a HelmRepository

This is an example that prepares the helm repository:
```python
from flux_local.kustomize import Kustomize
from flux_local.helm import Helm
from flux_local.manifest import HelmRepository

kustomize = Kustomize.build(TESTDATA_DIR)
repos = await kustomize.grep("kind=^HelmRepository$").objects()
helm = Helm("/tmp/path/helm", "/tmp/path/cache")
for repo in repos:
  helm.add_repo(HelmRepository.parse_doc(repo))
await helm.update()
```

Then to actually instantiate a template from a HelmRelease:
```python
from flux_local.manifest import HelmRelease

releases = await kustomize.grep("kind=^HelmRelease$").objects()
if not len(releases) == 1:
    raise ValueError("Expected only one HelmRelease")
tmpl = helm.template(
    HelmRelease.parse_doc(releases[0]),
    releases[0]["spec"].get("values"))
objects = await tmpl.objects()
for object in objects:
    print(f"Found object {object['apiVersion']} {object['kind']}")
```
"""

import datetime
import logging
from pathlib import Path
from typing import Any

import aiofiles
import yaml

from . import command
from .kustomize import Kustomize
from .manifest import HelmRelease, HelmRepository, CRD_KIND, SECRET_KIND
from .exceptions import HelmException

__all__ = [
    "Helm",
]

_LOGGER = logging.getLogger(__name__)


HELM_BIN = "helm"


class RepositoryConfig:
    """Generates a helm repository configuration from flux HelmRepository objects."""

    def __init__(self, repos: list[HelmRepository]) -> None:
        """Initialize RepositoryConfig."""
        self._repos = repos

    @property
    def config(self) -> dict[str, Any]:
        """Return a synthetic repository config object."""
        now = datetime.datetime.now(datetime.timezone.utc).replace(microsecond=0)
        return {
            "apiVersion": "",
            "generated": now.isoformat(),
            "repositories": [
                {
                    "name": f"{repo.namespace}-{repo.name}",
                    "url": repo.url,
                }
                for repo in self._repos
            ],
        }


class Helm:
    """Manages local HelmRepository state."""

    def __init__(self, tmp_dir: Path, cache_dir: Path) -> None:
        """Initialize Helm."""
        self._tmp_dir = tmp_dir
        self._repo_config_file = self._tmp_dir / "repository-config.yaml"
        self._flags = [
            "--registry-config",
            "/dev/null",
            "--repository-cache",
            str(cache_dir),
            "--repository-config",
            str(self._repo_config_file),
        ]
        self._repos: list[HelmRepository] = []

    def add_repo(self, repo: HelmRepository) -> None:
        """Add the specified HelmRepository to the local config."""
        self._repos.append(repo)

    def add_repos(self, repos: list[HelmRepository]) -> None:
        """Add the specified HelmRepository to the local config."""
        for repo in repos:
            self._repos.append(repo)

    async def update(self) -> None:
        """Return command line arguments to update the local repo.

        Typically the repository must be updated before doing any chart templating.
        """
        _LOGGER.debug("Updating %d repositories", len(self._repos))
        content = yaml.dump(RepositoryConfig(self._repos).config, sort_keys=False)
        async with aiofiles.open(str(self._repo_config_file), mode="w") as config_file:
            await config_file.write(content)
        await command.run(
            command.Command(
                [HELM_BIN, "repo", "update"] + self._flags, exc=HelmException
            )
        )

    async def template(
        self,
        release: HelmRelease,
        values: dict[str, Any] | None = None,
        skip_crds: bool = True,
        skip_tests: bool = True,
        skip_secrets: bool = False,
        kube_version: str | None = None,
        api_versions: str | None = None,
    ) -> Kustomize:
        """Return command line arguments to template the specified chart.

        The default values will come from the `HelmRelease`, though you can
        also specify values directory if not present in cluster manifest
        e.g. it came from a truncated yaml.
        """
        args: list[str] = [
            HELM_BIN,
            "template",
            release.name,
            release.chart.chart_name,
            "--namespace",
            release.namespace,
        ]
        if skip_crds:
            args.append("--skip-crds")
        if skip_tests:
            args.append("--skip-tests")
        if release.chart.version:
            args.extend(
                [
                    "--version",
                    release.chart.version,
                ]
            )
        if kube_version:
            args.extend(["--kube-version", kube_version])
        if api_versions:
            args.extend(["--api-versions", api_versions])
        if release.values and not values:
            values = release.values
        if values:
            values_path = self._tmp_dir / f"{release.release_name}-values.yaml"
            async with aiofiles.open(values_path, mode="w") as values_file:
                await values_file.write(yaml.dump(values, sort_keys=False))
            args.extend(["--values", str(values_path)])
        cmd = Kustomize([command.Command(args + self._flags, exc=HelmException)])
        skips = []
        if skip_crds:
            skips.append(CRD_KIND)
        if skip_secrets:
            skips.append(SECRET_KIND)
        if skips:
            cmd = cmd.skip_resources(skips)
        return cmd
