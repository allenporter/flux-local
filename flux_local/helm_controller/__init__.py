"""Helm controller package.

This package contains the implementation of the HelmRelease controller,
which manages the reconciliation of HelmRelease resources.
"""

from .controller import HelmReleaseController
from .artifact import HelmReleaseArtifact

__all__ = ["HelmReleaseController", "HelmReleaseArtifact"]
