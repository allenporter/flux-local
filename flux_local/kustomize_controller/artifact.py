"""Artifact types for the Kustomize Controller."""

from dataclasses import dataclass, field
from typing import Any

from flux_local.store.artifact import Artifact


@dataclass(frozen=True, kw_only=True)
class KustomizationArtifact(Artifact):
    """Artifact representing the result of building a Kustomization.

    Attributes:
        path: Path to the source directory containing kustomization.yaml
        manifests: List of rendered Kubernetes manifests
        revision: Source revision (commit SHA) if applicable
    """

    path: str
    manifests: list[dict[str, Any]] = field(default_factory=list)
    revision: str | None = None
