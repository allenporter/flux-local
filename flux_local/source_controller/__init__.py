"""The source controller module.

This module provides a controller for managing GitRepository and OCIRepository
resources.
"""

from .controller import SourceController
from .artifact import GitArtifact, OCIArtifact

__all__ = [
    "SourceController",
    "GitArtifact",
    "OCIArtifact",
]
