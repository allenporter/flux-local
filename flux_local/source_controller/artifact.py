"""Artifact representation."""

from dataclasses import dataclass

from flux_local.store.artifact import Artifact
from flux_local.manifest import OCIRepositoryRef, GitRepositoryRef


@dataclass(frozen=True, kw_only=True)
class GitArtifact(Artifact):
    """Git artifact.

    This object is written to the store when the GitRepository is reconciled.
    The path references a local filesystem path at the specified ref of the git
    repo.
    """

    url: str
    """URL of the git repository, for informational/logging purposes."""

    local_path: str
    """Local filesystem path to the git repository."""

    ref: GitRepositoryRef | None = None
    """Information about the version of the git repository."""


@dataclass(frozen=True, kw_only=True)
class OCIArtifact(Artifact):
    """OCI artifact.

    This object is written when the OCIRepository is reconciled. The path
    references a local filesystem path at the specified ref of the OCI repo.
    """

    url: str
    """URL of the oci repository, for informational/logging purposes."""

    local_path: str
    """Local filesystem path to the OCI repository."""

    ref: OCIRepositoryRef | None = None
    """Information about the version of the OCI repository."""
