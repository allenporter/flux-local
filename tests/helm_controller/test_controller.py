"""Tests for the helm controller."""

import asyncio
from collections.abc import Generator, AsyncGenerator
import tempfile
from pathlib import Path

import pytest
import yaml
import git
from syrupy.assertion import SnapshotAssertion

from flux_local.manifest import GitRepository, GIT_REPOSITORY
from flux_local.store.in_memory import InMemoryStore
from flux_local.store import Store, Status
from flux_local.helm_controller.artifact import HelmReleaseArtifact
from flux_local.helm_controller import HelmReleaseController
from flux_local.manifest import (
    NamedResource,
    HelmRelease,
    ConfigMap,
    Secret,
)
from flux_local.source_controller import GitArtifact
from flux_local.helm import Helm
from flux_local.task import get_task_service, TaskService, task_service_context


@pytest.fixture(name="task_service", autouse=True)
def task_service_fixture() -> Generator[TaskService, None, None]:
    """Create a task service for testing."""
    with task_service_context() as service:
        yield service


@pytest.fixture(name="tmp_dir")
def tmp_dir_fixture() -> Generator[Path, None, None]:
    """Create a temporary directory for test resources."""
    with tempfile.TemporaryDirectory() as td:
        yield Path(td)


@pytest.fixture(name="git_repo_tmp_dir", scope="module")
def git_repo_tmp_dir_fixture() -> Generator[Path, None, None]:
    """Create a temporary directory for test resources."""
    with tempfile.TemporaryDirectory() as td:
        yield Path(td)


@pytest.fixture(name="git_repo_dir", scope="module")
def git_repo_dir_fixture(git_repo_tmp_dir: Path) -> Path:
    """Create a local git repository for testing."""
    repo_path = git_repo_tmp_dir / "git-repo"
    repo_path.mkdir()

    # Initialize git repository
    repo = git.Repo.init(repo_path)
    repo.config_writer().set_value("user", "name", "myusername").release()
    repo.config_writer().set_value("user", "email", "myemail").release()

    # Create a test file
    test_file = repo_path / "test.txt"
    test_file.write_text("Test file content")

    # Add and commit the file
    repo.git.add(".")
    repo.git.commit(m="Initial commit")

    # Create a tag
    repo.git.tag("v1.0.0", m="Initial release")

    return repo_path


@pytest.fixture(name="helm_chart_git_path", scope="module")
def helm_chart_git_pathfixture(git_repo_dir: Path) -> Path:
    """Create a Helm repository directory for testing and return its path."""
    helm_chart_git_path = git_repo_dir / "helm-charts"
    helm_chart_git_path.mkdir(parents=True, exist_ok=True)

    # Create a test Helm chart
    chart_dir = helm_chart_git_path / "nginx"
    chart_dir.mkdir(exist_ok=True)

    # Create a basic Chart.yaml
    chart_yaml = chart_dir / "Chart.yaml"
    chart_yaml.write_text(
        """
apiVersion: v2
name: nginx
version: 1.0.0
appVersion: 1.14.2
""",
        encoding="utf-8",
    )

    # Add and commit the chart files
    repo = git.Repo(str(git_repo_dir))
    repo.git.add(".")
    repo.git.commit(m="Add Helm chart")

    return helm_chart_git_path


@pytest.fixture(name="chart_path", scope="module")
def chart_path_fixture(helm_chart_git_path: Path) -> Path:
    """Create a Helm chart directory with a Chart.yaml and return its path."""
    chart_dir = helm_chart_git_path / "nginx"
    chart_dir.mkdir(exist_ok=True)

    # Create a basic Chart.yaml
    chart_yaml = chart_dir / "Chart.yaml"
    chart_yaml.write_text(
        """
apiVersion: v2
name: nginx
description: A Helm chart for nginx
version: 1.0.0
appVersion: 1.14.2
""",
        encoding="utf-8",
    )

    # Create a basic values.yaml
    values_yaml = chart_dir / "values.yaml"
    values_yaml.write_text(
        """
replicaCount: 7
""",
        encoding="utf-8",
    )
    templates = chart_dir / "templates"
    templates.mkdir(exist_ok=True)

    config_map_yaml = templates / "configmap.yaml"
    config_map_yaml.write_text(
        """
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-configmap
data:
  myvalue: "Hello World with {{ .Values.replicaCount }} replicas"
""",
        encoding="utf-8",
    )

    return chart_dir


@pytest.fixture(name="config_map")
def config_map_fixture() -> ConfigMap:
    """Create a test ConfigMap."""
    config_map_yaml = """
apiVersion: v1
kind: ConfigMap
metadata:
  name: test-config
  namespace: test-ns
  labels:
    flux_local_test: "true"
data:
  values.yaml: |
    replicaCount: 2
"""
    config_map = ConfigMap.parse_doc(yaml.safe_load(config_map_yaml))
    return config_map


@pytest.fixture(name="secret")
def secret_fixture() -> Secret:
    """Create a test Secret."""
    secret_yaml = """
apiVersion: v1
kind: Secret
metadata:
  name: test-secret
  namespace: test-ns
  labels:
    flux_local_test: "true"
type: Opaque
stringData:
  values.yaml: |
    replicaCount: 3
"""
    secret = Secret.parse_doc(yaml.safe_load(secret_yaml))
    return secret


@pytest.fixture(name="helm_release")
def helm_release_fixture(
    git_repo_dir: Path,
    chart_path: Path,
    helm_chart_git_path: Path,
    config_map: ConfigMap,
    secret: Secret,
) -> HelmRelease:
    """Create a test HelmRelease."""
    chart_name = chart_path.relative_to(git_repo_dir)
    helm_release_yaml = f"""
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: test-release
  namespace: test-ns
spec:
  chart:
    spec:
      chart: {chart_name}
      version: 1.0.0
      sourceRef:
        kind: GitRepository
        name: test-repo
        namespace: test-ns
        ref:
          tag: v1.0.0
  valuesFrom:
    - kind: ConfigMap
      name: {config_map.name}
      valuesKey: values.yaml
      targetPath: .
    - kind: Secret
      name: {secret.name}
      valuesKey: values.yaml
      targetPath: .
"""
    helm_release = HelmRelease.parse_doc(yaml.safe_load(helm_release_yaml))
    return helm_release


@pytest.fixture(name="git_repo")
def git_repo_fixture() -> GitRepository:
    """Create a test GitRepository."""
    git_repo_yaml = """
apiVersion: source.toolkit.fluxcd.io/v1
kind: GitRepository
metadata:
  name: test-repo
  namespace: test-ns
spec:
  url: https://github.com/example/helm-charts
  ref:
    tag: v1.0.0
  interval: 1m0s
"""
    git_repo = GitRepository.parse_doc(yaml.safe_load(git_repo_yaml))
    return git_repo


@pytest.fixture(name="store")
def store_fixture() -> Store:
    """Create a test store."""
    return InMemoryStore()


@pytest.fixture(name="helm")
def helm_fixture(tmp_dir: Path) -> Helm:
    """Create a test Helm instance."""
    cache_dir = tmp_dir / "helm-cache"
    cache_dir.mkdir(exist_ok=True)
    return Helm(tmp_dir=tmp_dir, cache_dir=cache_dir)


@pytest.fixture(name="controller")
async def controller_fixture(
    store: Store, helm: Helm
) -> AsyncGenerator[HelmReleaseController, None]:
    """Create a test controller."""
    controller = HelmReleaseController(store, helm)
    yield controller
    await controller.close()


async def test_helm_release_reconciliation(
    store: Store,
    controller: HelmReleaseController,
    helm_release: HelmRelease,
    git_repo: GitRepository,
    helm_chart_git_path: Path,
    chart_path: Path,
    config_map: ConfigMap,
    secret: Secret,
    git_repo_dir: Path,
    snapshot: SnapshotAssertion,
) -> None:
    """Test basic HelmRelease reconciliation with dependencies."""
    # Add the ConfigMap and Secret resources first
    store.add_object(config_map)
    store.add_object(secret)

    # Add the GitRepository
    git_repo_rid = NamedResource(git_repo.kind, git_repo.namespace, git_repo.name)
    store.add_object(git_repo)

    # Set the artifact for the GitRepository
    repo_artifact = GitArtifact(
        url=git_repo.url,
        local_path=str(git_repo_dir),
        ref=git_repo.ref,
    )
    store.set_artifact(git_repo_rid, repo_artifact)

    # Update status to READY
    store.update_status(git_repo_rid, Status.READY)

    helm_release_rid = NamedResource(
        helm_release.kind, helm_release.namespace, helm_release.name
    )

    # Add the HelmRelease to trigger reconciliation
    store.add_object(helm_release)

    task_service = get_task_service()
    await task_service.block_till_done()
    assert not task_service.get_num_active_tasks()

    status = store.get_status(helm_release_rid)
    assert status is not None
    assert status.status == Status.READY

    artifact = store.get_artifact(helm_release_rid, HelmReleaseArtifact)
    assert artifact is not None
    assert artifact.chart_name == helm_release.chart.chart_name

    # Verify objects are applied
    objects = store.list_objects()
    assert [
        obj
        for obj in objects
        # GitRepository has a tmpdir random path so ignore
        if hasattr(obj, "kind") and obj.kind != "GitRepository"
    ] == snapshot


async def test_helm_release_becomes_ready_after_gitrepo_ready(
    store: Store,
    controller: HelmReleaseController,
    helm_release: HelmRelease,
    git_repo: GitRepository,
    config_map: ConfigMap,
    secret: Secret,
    git_repo_dir: Path,
    chart_path: Path,
    helm_chart_git_path: Path,
) -> None:
    """Test HelmRelease becomes ready after GitRepository becomes ready."""
    # Add the ConfigMap and Secret resources first
    store.add_object(config_map)
    store.add_object(secret)

    # Add the GitRepository but don't mark it as ready yet
    git_repo_rid = NamedResource(GIT_REPOSITORY, git_repo.namespace, git_repo.name)
    store.add_object(git_repo)

    # Add the HelmRelease to trigger reconciliation
    helm_release_rid = NamedResource(
        helm_release.kind, helm_release.namespace, helm_release.name
    )
    store.add_object(helm_release)

    # Wait for reconciliation to start
    await asyncio.sleep(0.001)

    repo_artifact = GitArtifact(
        url=git_repo.url,
        local_path=str(git_repo_dir),
        ref=git_repo.ref,
    )
    store.set_artifact(git_repo_rid, repo_artifact)
    store.update_status(git_repo_rid, Status.READY)

    task_service = get_task_service()
    await task_service.block_till_done()
    assert not task_service.get_num_active_tasks()

    # Verify the HelmRelease is ready
    status = store.get_status(helm_release_rid)
    assert status is not None
    assert status.status == Status.READY

    # Verify the artifact is set
    artifact = store.get_artifact(helm_release_rid, HelmReleaseArtifact)
    assert artifact is not None
    assert artifact.chart_name == helm_release.chart.chart_name


async def test_helm_release_fails_with_missing_dependency(
    store: Store,
    controller: HelmReleaseController,
    helm_release: HelmRelease,
    git_repo: GitRepository,
    helm_chart_git_path: Path,
    chart_path: Path,
    git_repo_dir: Path,
) -> None:
    """Test HelmRelease fails when a dependency is missing."""
    # Add the GitRepository
    git_repo_rid = NamedResource(git_repo.kind, git_repo.namespace, git_repo.name)
    store.add_object(git_repo)

    # Set the artifact for the GitRepository
    repo_artifact = GitArtifact(
        url=git_repo.url,
        local_path=str(git_repo_dir),
        ref=git_repo.ref,
    )
    store.set_artifact(git_repo_rid, repo_artifact)

    # Update status to READY
    store.update_status(git_repo_rid, Status.READY)

    # Add the HelmRelease to trigger reconciliation
    store.add_object(helm_release)

    # Wait for reconciliation to complete and fail
    task_service = get_task_service()
    await task_service.block_till_done()
    assert not task_service.get_num_active_tasks()

    status = store.get_status(
        NamedResource(helm_release.kind, helm_release.namespace, helm_release.name)
    )
    assert status is not None
    assert status.status == Status.FAILED
    # TODO: Make this fail or pending without a timeout
    assert status.error == "Reconciliation failed: TimeoutError: "


async def test_helm_release_fails_failed_dependency(
    store: Store,
    controller: HelmReleaseController,
    helm_release: HelmRelease,
    git_repo: GitRepository,
    helm_chart_git_path: Path,
    chart_path: Path,
    git_repo_dir: Path,
    config_map: ConfigMap,
    secret: Secret,
) -> None:
    """Test HelmRelease fails when a dependency is missing."""
    # Add the ConfigMap and Secret resources first
    store.add_object(config_map)
    store.add_object(secret)

    # Add the GitRepository
    git_repo_rid = NamedResource(git_repo.kind, git_repo.namespace, git_repo.name)
    store.add_object(git_repo)

    store.update_status(git_repo_rid, Status.FAILED, "Some error")

    # Add the HelmRelease to trigger reconciliation
    store.add_object(helm_release)

    task_service = get_task_service()
    await task_service.block_till_done()
    assert not task_service.get_num_active_tasks()

    status = store.get_status(
        NamedResource(helm_release.kind, helm_release.namespace, helm_release.name)
    )
    assert status is not None
    assert status.status == Status.FAILED
    assert (
        status.error
        == "Reconciliation failed: HelmControllerException: Dependency GitRepository/test-ns/test-repo: Failed: Some error"
    )


async def test_helm_release_fails_with_nonexistent_chart(
    store: Store,
    controller: HelmReleaseController,
    helm_release: HelmRelease,
    git_repo: GitRepository,
    helm_chart_git_path: Path,
    chart_path: Path,
    git_repo_dir: Path,
    config_map: ConfigMap,
    secret: Secret,
) -> None:
    """Test HelmRelease fails when chart directory doesn't exist in GitRepository."""
    # Add the ConfigMap and Secret resources first
    store.add_object(config_map)
    store.add_object(secret)

    # Create a HelmRelease that points to a non-existent chart directory
    helm_release.chart.name = "non-existent-chart"
    helm_release.chart.repo_name = git_repo.name
    helm_release.chart.repo_namespace = git_repo.namespace
    helm_release.chart.repo_kind = git_repo.kind

    # Add the GitRepository
    git_repo_rid = NamedResource(git_repo.kind, git_repo.namespace, git_repo.name)
    store.add_object(git_repo)

    # Set the artifact for the GitRepository
    repo_artifact = GitArtifact(
        url=git_repo.url,
        local_path=str(git_repo_dir),
        ref=git_repo.ref,
    )
    store.set_artifact(git_repo_rid, repo_artifact)

    # Update status to READY
    store.update_status(git_repo_rid, Status.READY)

    # Add the HelmRelease to trigger reconciliation
    store.add_object(helm_release)

    task_service = get_task_service()
    await task_service.block_till_done()
    assert not task_service.get_num_active_tasks()

    status = store.get_status(
        NamedResource(helm_release.kind, helm_release.namespace, helm_release.name)
    )
    assert status is not None
    assert status.status == Status.FAILED
    assert status.error is not None
    assert "not found" in status.error
