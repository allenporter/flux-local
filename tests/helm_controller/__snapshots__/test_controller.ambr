# serializer version: 1
# name: test_helm_release_reconciliation
  list([
    ConfigMap(name='test-config', namespace='test-ns', data={'values.yaml': 'replicaCount: 2\n'}, binary_data=None),
    Secret(name='test-secret', namespace='test-ns', data=None, string_data={'values.yaml': '..PLACEHOLDER_values.yaml..'}),
    HelmRelease(name='test-release', namespace='test-ns', chart=HelmChart(name='helm-charts/nginx', version='1.0.0', repo_name='test-repo', repo_namespace='test-ns', repo_kind='GitRepository'), target_namespace=None, values=None, values_from=[ValuesReference(kind='ConfigMap', name='test-config', values_key='values.yaml', target_path='.', optional=False), ValuesReference(kind='Secret', name='test-secret', values_key='values.yaml', target_path='.', optional=False)], images=None, labels=None, disable_schema_validation=False, disable_openapi_validation=False),
    ConfigMap(name='test-release-configmap', namespace=None, data={'myvalue': 'Hello World with 7 replicas'}, binary_data=None),
  ])
# ---
